core/vm/actor/builtin/market/actor.cpp

```diff
@@ -5,7 +5,434 @@
+      OUTCOME_TRY(set->flush());
+      OUTCOME_TRY(deals_by_party.set(*address, *set));
+    }
+    return outcome::success();
+  }
+
+  TokenAmount clientPayment(ChainEpoch epoch,
+                            const DealProposal &deal,
+                            const DealState &deal_state) {
+    auto end = std::min(epoch, deal.end_epoch);
+    if (deal_state.slash_epoch != kChainEpochUndefined) {
+      end = std::min(end, deal_state.slash_epoch);
+    }
+    auto start = deal.start_epoch;
+    if (deal_state.last_updated_epoch != kChainEpochUndefined) {
+      start = std::max(start, deal_state.last_updated_epoch);
+    }
+    return deal.storage_price_per_epoch * (end - start);
+  }
+
+  outcome::result<State> loadState(Runtime &runtime) {
+    OUTCOME_TRY(state, runtime.getCurrentActorStateCbor<State>());
+    state.load(runtime);
+    return std::move(state);
+  }
+
+  outcome::result<void> commitState(Runtime &runtime, State &state) {
+    OUTCOME_TRY(state.flush());
+    return runtime.commitState(state);
+  }
+
+  outcome::result<void> assertCallerIsMiner(Runtime &runtime) {
+    OUTCOME_TRY(code, runtime.getActorCodeID(runtime.getImmediateCaller()));
+    if (code != kStorageMinerCodeCid) {
+      return VMExitCode::MARKET_ACTOR_WRONG_CALLER;
+    }
+    return outcome::success();
+  }
+
+  outcome::result<void> assertCallerIsSignable(Runtime &runtime) {
+    OUTCOME_TRY(code, runtime.getActorCodeID(runtime.getImmediateCaller()));
+    if (!isSignableActor(code)) {
+      return VMExitCode::MARKET_ACTOR_WRONG_CALLER;
+    }
+    return outcome::success();
+  }
+
+  outcome::result<std::pair<Address, Address>> escrowAddress(
+      Runtime &runtime, const Address &address) {
+    OUTCOME_TRY(nominal, runtime.resolveAddress(address));
+    OUTCOME_TRY(code, runtime.getActorCodeID(nominal));
+    if (code == kStorageMinerCodeCid) {
+      OUTCOME_TRY(miner, requestMinerControlAddress(runtime, nominal));
+      if (runtime.getImmediateCaller() != miner.worker
+          && runtime.getImmediateCaller() != miner.owner) {
+        return VMExitCode::MARKET_ACTOR_WRONG_CALLER;
+      }
+      return std::make_pair(nominal, miner.owner);
+    }
+    if (!isSignableActor(code)) {
+      return VMExitCode::MARKET_ACTOR_WRONG_CALLER;
+    }
+    return std::make_pair(nominal, nominal);
+  }
+
+  outcome::result<void> unlockBalance(State &state,
+                                      const Address &address,
+                                      TokenAmount amount) {
+    if (amount < 0) {
+      return VMExitCode::MARKET_ACTOR_ILLEGAL_STATE;
+    }
+    return state.locked_table.subtract(address, amount);
+  }
+
+  outcome::result<void> slashBalance(State &state,
+                                     const Address &address,
+                                     TokenAmount amount) {
+    if (amount < 0) {
+      return VMExitCode::MARKET_ACTOR_ILLEGAL_STATE;
+    }
+    OUTCOME_TRY(state.escrow_table.subtract(address, amount));
```

Comment:
In the reference implementation, the amount be subtracted from the locked table only

Reply:
No, both are subtracted, [market_state.go](https://github.com/filecoin-project/specs-actors/blob/master/actors/builtin/market/market_state.go#L323)